### ISSUE TITLE
[TvOS] Memory Leak Using autoFocus Feature with TVFocusGuideView

### ISSUE THREAD

**Issue by DBarbosaDev** [react-native-tvos/react-native-tvos#506]
### Description

While conducting tests on the new TVFocusGuideView features - requestTVFocus, autoFocus, and setDestinations, a memory leak issue was discovered.

The investigation points to the autoFocus feature as the cause of this memory leak. To verify this, we developed a proof-of-concept (POC) that encapsulated the logic of the TVFocusGuideView and its features independently.

The memory usage escalates when the TVFocusGuideView component is repeatedly destroyed and recreated with nested elements such as flatLists, other TVFocusGuideViews, TouchableOpacity’s, etc. 
Note that a direct correlation between the number of child elements and the amount of memory leakage is observerd.

In addition, a specific scenario stood out: When we use valid destinations (either via the destination property or using the setDestinations function) in conjunction with the autoFocus feature, the memory leak ceased to occur.

Check the following video:

https://github.com/react-native-tvos/react-native-tvos/assets/48337494/64bb68b8-4fa7-4387-a9ff-7ca14e661626 

### Version

0.69.8-1 and 0.71.8-0rc0

### Output of `npx react-native info`

**OS: macOS 13.1**

   CPU: (8) x64 Intel(R) Core(TM) i5-8257U CPU @ 1.40GHz
    Memory: 748.06 MB / 16.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 14.17.0 - /usr/local/bin/node
    Yarn: 1.22.10 - /usr/local/bin/yarn
    npm: 7.14.0 - /usr/local/bin/npm
    Watchman: 4.9.0 - /usr/local/bin/watchman
  Managers:
    CocoaPods: 1.11.3 - /usr/local/bin/pod
  SDKs:
    iOS SDK:
      Platforms: DriverKit 22.4, iOS 16.4, macOS 13.3, tvOS 16.4, watchOS 9.4
    Android SDK: Not Found
  IDEs:
    Android Studio: 4.2 AI-202.7660.26.42.7351085
    Xcode: 14.3/14E222b - /usr/bin/xcodebuild
  Languages:
    Java: 16.0.1 - /usr/bin/javac
  npmPackages:
    @react-native-community/cli: Not Found
    react: ^18.2.0 => 18.2.0 
    react-native: Not Found
    react-native-macos: Not Found
    react-native-tvos:  0.71.8-0rc0 
  npmGlobalPackages:
    *react-native*: Not Found

**TVOS:   tvOS 16.4(20L497)**

	Apple TV 4K (1.ª generation)
	Model number: A1842
	Capacity: 32 GB
	HDMI 2.0a (2160p, Dolby Vision, and HDR10)

### Steps to reproduce

**Pre-conditions**

- A component screen must be present;
- The screen should be class-based.

**Steps to reproduce**

1. Create state **A** to control rendering;
2. Add the TVFocusGuideView component with autoFocus activated;
3. Render the component created in step2 conditionally, based on whether state A is true;
4. Within the TVFocusGuideView, add for instance, 3 flatlists with 5 touchables each;
5. Set up a handler to intercept MENU or BACK button click events from the remote;
6. Inside this handler, toggle the value of state A;
7. Press the button multiple times defined in step 5;
8. Observe the results.

### Snack, code example, screenshot, or link to a repository

```javascript
import React, { PureComponent } from 'react';
import { View, StyleSheet, FlatList, Text, TouchableOpacity, TVFocusGuideView, FocusGuideMethods } from 'react-native';
import { HWEventHandler, HWEventType } from '../common/hardwareEvents';

const styles = StyleSheet.create({
  container: {
    backgroundColor: '#E0E0E0',
    flex: 1,
  },
  button: {
    backgroundColor: '#404040',
    marginHorizontal: 10,
    width: 200,
    alignItems: 'center',
    justifyContent: 'center',
  },
  headerItem: {
    backgroundColor: '#404040',
    marginHorizontal: 10,
    width: 310,
    alignItems: 'center',
    justifyContent: 'center',
  },
  card: {
    backgroundColor: '#404040',
    marginHorizontal: 10,
    width: 200,
    alignItems: 'center',
    justifyContent: 'center',
  },
  tvFocusGuide: {
    width: '100%',
    height: '100%',
  },
  tvFocusGuideRail: {
    position: 'absolute',
    top: 300,
    left: 100,
    width: '100%',
  },
  flatlist: {
    height: 100,
  },
  aux: {
    backgroundColor: '#FF0000',
    height: 1,
    width: 1,
  },
});

const data = [
  {
    key: 0,
  },
  {
    key: 1,
  },
  {
    key: 2,
  },
  {
    key: 3,
  },
  {
    key: 4,
  },
  {
    key: 5,
  },
  {
    key: 6,
  },
  {
    key: 7,
  },
  {
    key: 8,
  },
  {
    key: 9,
  },
  {
    key: 10,
  },
  {
    key: 11,
  },
  {
    key: 12,
  },
  {
    key: 13,
  },
  {
    key: 14,
  },
  {
    key: 15,
  },
  {
    key: 16,
  },
  {
    key: 17,
  },
  {
    key: 18,
  },
  {
    key: 19,
  },
  {
    key: 20,
  },
  {
    key: 21,
  },
  {
    key: 22,
  },
  {
    key: 23,
  },
  {
    key: 24,
  },
  {
    key: 25,
  },
  {
    key: 26,
  },
  {
    key: 27,
  },
  {
    key: 28,
  },
  {
    key: 29,
  },
  {
    key: 30,
  },
  {
    key: 31,
  },
  {
    key: 32,
  },
  {
    key: 33,
  },
  {
    key: 34,
  },
  {
    key: 35,
  },
  {
    key: 36,
  },
  {
    key: 37,
  },
  {
    key: 38,
  },
  {
    key: 39,
  },
];

interface State {
  render: boolean;
  destinationItemRef: TouchableOpacity | null;
}

class FocusTestScreen extends PureComponent<unknown, State> {
  tvFocusGuideRef: React.RefObject<View & FocusGuideMethods> = React.createRef<View & FocusGuideMethods>();
  destinationItemRef: TouchableOpacity | null = null;
  eventHandler: HWEventHandler;

  // -------
  isAutoFocusOn = true;
  toUseDestinations = false;
  // -------

  constructor(props: any) {
    super(props);

    this.state = {
      render: true,
      destinationItemRef: null,
    };

    this.eventHandler = HWEventHandler.bind(this);
  }

  componentDidMount(): void {
    this.eventHandler.addEventListener(HWEventType.BACK_MENU, () => {
      this.setState((prevState) => {
        console.log('BACK CLICKED - TVFocusGuideView rendered? [', !prevState.render, ']');
        console.log('--');
        return {
          render: !prevState.render,
        };
      });
    });
  }

  componentDidUpdate(_prevProps: Readonly<unknown>, prevState: Readonly<State>): void {
    const { destinationItemRef } = this.state;

    if (destinationItemRef !== prevState.destinationItemRef && this.toUseDestinations) {
      this.tvFocusGuideRef.current?.setDestinations([destinationItemRef]);
    }
  }

  setDestination = (ref: TouchableOpacity | null) => {
    if (ref) {
      this.setState({ destinationItemRef: ref });
    }
  };

  renderFlatlist = ({ item, index }: { item: any; index: number }): JSX.Element | null => {
    return (
      <TouchableOpacity style={styles.card} ref={this.setDestination}>
        <Text>{item.key}</Text>
      </TouchableOpacity>
    );
  };

  renderHeader = ({ item }: { item: any }): JSX.Element | null => {
    return (
      <TouchableOpacity style={styles.headerItem}>
        <Text>{item.key}</Text>
      </TouchableOpacity>
    );
  };

  renderButtons = ({ item }: { item: any }): JSX.Element | null => {
    return (
      <TouchableOpacity style={styles.button}>
        <Text>{item.key}</Text>
      </TouchableOpacity>
    );
  };

  render(): JSX.Element {
    const { render } = this.state;

    return (
      <View style={styles.container}>
        <TouchableOpacity style={styles.aux} />
        {render ? (
          <TVFocusGuideView ref={this.tvFocusGuideRef} style={styles.tvFocusGuide} autoFocus={this.isAutoFocusOn}>
            <TVFocusGuideView style={styles.tvFocusGuideRail}>
              <FlatList
                windowSize={2}
                data={data}
                style={styles.flatlist}
                horizontal
                renderItem={this.renderFlatlist}
              />
              <FlatList
                windowSize={2}
                data={data}
                style={styles.flatlist}
                horizontal
                renderItem={this.renderFlatlist}
              />
              <FlatList
                windowSize={2}
                data={data}
                style={styles.flatlist}
                horizontal
                renderItem={this.renderFlatlist}
              />
              <FlatList
                windowSize={2}
                data={data}
                style={styles.flatlist}
                horizontal
                renderItem={this.renderFlatlist}
              />
              <FlatList
                windowSize={2}
                data={data}
                style={styles.flatlist}
                horizontal
                renderItem={this.renderFlatlist}
              />
              <FlatList
                windowSize={2}
                data={data}
                style={styles.flatlist}
                horizontal
                renderItem={this.renderFlatlist}
              />
            </TVFocusGuideView>
          </TVFocusGuideView>
        ) : null}
      </View>
    );
  }
}

export default FocusTestScreen;

```

**Comment by yusufyildirim** [comment]
Thanks for reporting  @DBarbosaDev! I'll try to check this one out until next week 👍

**Comment by yusufyildirim** [comment]
@DBarbosaDev I checked it yesterday and identified the point that causes the problem but I don't have a fix yet. I have other stuff at work to sort out first so I'll get back to this ASAP. Just wanted to give an update

**Comment by DBarbosaDev** [comment]
Hello @yusufyildirim do you have any update?

**Comment by yusufyildirim** [comment]
Hi @DBarbosaDev. I was on vacation and couldn't find time to continue on this after I came back. I will get this done in the next couple of days.

**Comment by yusufyildirim** [comment]
PR is ready https://github.com/react-native-tvos/react-native-tvos/pull/525

**Comment by DBarbosaDev** [comment]
Thank you very much for your effort!

**Comment by douglowder** [comment]
This is merged and released in both 0.71 and 0.72 latest releases, closing.
